### ISSUE TITLE
Input iterator's column data member doesn't account for multibyte characters (UTF-8)

### ISSUE THREAD

**Issue by catsidhe** [taocpp/PEGTL#369]
Unless I'm mistaken, `column` currently counts bytes. I ended up re-calculating the column by walking back from `data` until I see a newline character or reach the beginning of the content, counting characters while ignoring non-first UTF-8 bytes. Would've been nicer if it was handled by PEGTL.

**Comment by ColinH** [comment]
Am I right to assume that what you are counting are codepoints? That might be possible to add for PEGTL 4.0, though we have been apprehensive to go down that road because the next step would be to take composed characters and grapheme clusters into consideration. That's a big can of worms to open.

**Comment by catsidhe** [comment]
Yes, just the code points, bytes outside 0x80..0xbf range. I figured it'd cover the majority of use cases, while being simple and cheap.
```cpp
constexpr auto count_utf8(std::string_view sv) -> std::size_t {
  std::size_t count{};
  for (std::uint8_t byte: sv) if (byte <= 0x7f || byte >= 0xc0) ++count;
  return count;
}
```

**Comment by ColinH** [comment]
Are you using lazy or eager position tracking?

**Comment by catsidhe** [comment]
Sorry, I don't quite know how to answer that. Maybe because I use it in a tree? Probably incorrectly, too!
```cpp
auto from_utf8(std::string_view) -> std::u32string;
inline auto line_width(std::u32string_view sv) -> int {
  return std::accumulate(sv.begin(), sv.end(), 0, [](int w, char32_t uc) {
    return w + std::max(width(uc), 0);
  });
}
class node {
  // ...
  template<typename _rule, typename _input, typename... _states>
  void start(_input const &in, _states &&...) noexcept {
    auto i = in.iterator();
    line = static_cast<std::uint32_t>(i.line);
    auto data = i.data - (i.column - 1); // Can I do this? Probably not.
    column = line_width(from_utf8({data, static_cast<std::size_t>(i.data - data)})) + 1;
  }
  // ...
};
```
Actually I'm no longer sure how to best calculate columns. Consider this snippet:
```cpp
int /*あいうえお*/ 123;
```
According to VSCode, the digit `1` is on column 15. It treats each character as occupying exactly one column.

Clang thinks it's on column 25 (counting bytes, like PEGTL):
```
test.cpp:1:25: error: expected unqualified-id
    1 | int /*あいうえお*/ 123;
      |                    ^
```
GCC says 20 (presumably using `wcwidth`):
```cpp
test.cpp:1:20: error: expected unqualified-id before numeric constant
   1 | int /*あいうえお*/ 123;
     |                    ^~~
```
(Arrows point to `1` in both examples above in the terminal.)

For the best flexibility, perhaps a user-supplied character traits class is needed, that gets to decide how many columns each code point takes.

**Comment by ColinH** [comment]
I've been keeping this in the back of my mind but don't see an easy solution within the PEGTL, not even with all the changes to the inputs and position tracking coming in version 4.0.0.

It might be done for lazy position tracking, which would be fine if it's only used for a `parse_error`, but would make things very slow (quadratic parsing time) when used with the `parse_tree`.

It would also make things more lopsided, and we would still only be counting code points, not "visible characters" which can be composed from multiple consecutive code points.

Doing this well needs Unicode normalization and/or some other Unicode deep dive that would make this library a lot bigger and/or complicated than it already is.

That would not necessarily deter me from trying, however I simply do not have the time to pursue this - it's already taking too much time to get version 4.0.0 done without it.

My apologies for closing this without a solution.
